metals/src/main/scala/scala/meta/internal/metals/debug/RuntimeSourceAdapter.scala

```diff
@@ -0,0 +1,59 @@
+package scala.meta.internal.metals.debug
+
+import ch.epfl.scala.bsp4j.BuildTargetIdentifier
+import org.eclipse.lsp4j.debug.Source
+import scala.meta.internal.metals.BuildTargets
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.WorkspaceSymbolProvider
+
+final class RuntimeSourceAdapter(
+    symbolProvider: WorkspaceSymbolProvider,
+    buildTargets: BuildTargets,
+    targets: List[BuildTargetIdentifier]
+) {
+  def adapt(source: Source): Boolean = {
+    val symbol = searchRealPath(source)
+
+    symbol match {
+      case Some(symbol) =>
+        source.setPath(symbol)
+        true
+      case None =>
+        false
+    }
+  }
+
+  private def searchRealPath(source: Source): Option[String] = {
+    if (source == null) None
+    else {
+      searchAsClassPathSymbol(source).orElse(searchAsSourceFile(source))
+    }
+  }
+
+  private def searchAsClassPathSymbol(source: Source): Option[String] = {
+    val container =
+      source.getPath.stripSuffix(source.getName).replaceAll("/", ".")
+    val pattern = source.getName
+      .stripSuffix(".scala")
+      .stripSuffix(".java")
+
+    val symbols = for {
+      target <- targets.view
+      symbol <- symbolProvider.searchExact(pattern, Some(target))
```

Comment:
It's quite heavyweight to go via `WorkspaceSymbolProvider` to find the location of a definition. Did you try using the `DefinitionIndex` directly? It's possible to compute mtags "toplevels" on the fly given the source file.

Reply:
WorkspaceSymbolProvider supports fuzzy symbol searching, which I don't think we need here

Reply:
You are right, using DefinitionProvider works! It has one regression in regards to the WorksapceSymbolProvider but I think it can be addressed later: ambiguous source files, like:
- `foo/a/b/Foo.scala`
- `bar/a/b/Foo.scala`
where `foo` and `bar` are different, unrelated modules (let's say with variants of classes for different scala versions or something like that).

In this case, the `OnDemandSymbolIndex` fails us, because the top-level symbols from the latter module would overwrite the symbols from the earlier

Reply:
The same limitation applies for “goto definition” and this issue is totally fixable.